python/cudf/cudf/core/index.py

```diff
@@ -1403,7 +1403,7 @@ def __repr__(self):
     @_cudf_nvtx_annotate
     def __getitem__(self, index):
         res = self._get_elements_from_column(index)
-        if not isinstance(index, int):
+        if isinstance(res, ColumnBase):
```

Comment:
Chasing the set of things that are `int`-like is always a bad idea, so instead dispatch on the complement (`_get_elements_from_column :: Scalar | Array -> Scalar | Column`)